### PR TITLE
Allows players to name seed packs to keep track of modified plants

### PR DESCRIPTION
**Pull request by theo-3**
## What Does This PR Do
Adds a 'variant' var to seed packs which can be set by players with a pen or with the dna modifier. This variable is shown in the pack's name, in the seed extractor, in the dna modifier and in hydroponic trays' examine. Extracting seed packs from named seeds will keep the name.

## Why It's Good For The Game
Helps keep track of what modified plants do, for example a tomato plant labeled 'healing' and another labeled 'acid', or a gaia plant labeled 'perrenial densified chemicals'.

## Images of changes

The gif doesn't show the dna modifier UI because I messed up the window capture. Sorry.
Capture1
Capture2
Capture1

## Changelog
:cl:
add: Adds a 'variant' var to seed packets which can be set by using a pen on seeds or a tray, or from the dna modifier. It shows up in the packet's name, in plant trays examine, in the seed extractor, in the dna modifier and in harvested plants. This is intended to help keep track of what modified plants do, such as wheat labeled 'healing'.
/:cl:

### CODE REVIEW

File: code/modules/hydroponics/gene_modder.dm [hunk 1]
```diff
@@ -391,6 +398,19 @@
 	else if(href_list["abort"])
 		operation = ""
 		target = null
+	else if(href_list["set_v"])
+		if(!seed)
+			return
+		var/V = sanitize(stripped_input(usr, "Choose variant name:", "Plant Variant Naming", seed.variant, MAX_MESSAGE_LEN))
```
Comment (dearmochi) [inline code/modules/hydroponics/gene_modder.dm:404]: Might want to use a max length lower than 1024, or you could end up with absurdly long names that fill up the chat. Perhaps 64 characters is saner.

File: code/modules/hydroponics/seeds.dm [hunk 2]
```diff
@@ -317,16 +320,32 @@
...
-	if (istype(O, /obj/item/plant_analyzer))
+	if(istype(O, /obj/item/plant_analyzer))
 		to_chat(user, "<span class='info'>*---------*\n This is \a <span class='name'>[src]</span>.</span>")
 		var/text = get_analyzer_text()
 		if(text)
 			to_chat(user, "<span class='notice'>[text]</span>")
 
 		return
+	if(istype(O, /obj/item/pen))
+		var/V = sanitize(stripped_input(user, "Choose variant name:", "Plant Variant Naming", variant, MAX_MESSAGE_LEN))
```
Comment (dearmochi) [inline code/modules/hydroponics/seeds.dm:331]: Ditto

File: code/modules/hydroponics/seeds.dm [hunk 3]
```diff
@@ -317,16 +320,32 @@
...
+		if(!Adjacent(user)) //prevent remote naming of seeds
+			return
+		if(V == "" && variant != "")
+			user.show_message("<span class='info'>You remove the variant designation from the [plantname].</span>")
+		else if(V != "")
+			user.show_message("<span class='info'>You designate the [plantname] as the [V] variant.</span>")
+		variant = V
+		apply_variant_name()
```
Comment (dearmochi) [inline code/modules/hydroponics/seeds.dm:339]:
```suggestion
		apply_variant_name()
		return
```

So it doesn't proceed to the default behaviour

File: code/modules/hydroponics/seeds.dm [hunk 4]
```diff
@@ -317,16 +320,32 @@
...
 	..() // Fallthrough to item/attackby() so that bags can pick seeds up
 
-
+/obj/item/seeds/proc/apply_variant_name()
+	var/P = findtext(name, "\[")
+	var/V = (variant == "") ? "" : (" \[" + variant + "]") // If we have a non-empty variant add it to the name
+	if (P != 0)
+		name = copytext(name, 1, P - 1) + V // If there was already a variant in the name remove it
```
Comment (dearmochi) [inline code/modules/hydroponics/seeds.dm:346]: Consider the `initial(varname)` proc to get the initial name of a datum, would be better than text manipulation

**Review by dearmochi — CHANGES_REQUESTED** [review]

File: code/modules/hydroponics/gene_modder.dm (hunk 1 above)
Comment (theo-3) [reply code/modules/hydroponics/gene_modder.dm]: Thanks, I copy pasted that from another file and didn't know what the max len did

File: code/modules/hydroponics/seeds.dm (hunk 4 above)
Comment (theo-3) [reply code/modules/hydroponics/seeds.dm:347]: Wouldn't that delete the 'experimental' from modified seeds?

Comment (dearmochi) [reply code/modules/hydroponics/seeds.dm:347]: Good point, this is fine then

**Comment by theo-3** [comment]
Thanks for the help!

**Review by dearmochi — APPROVED** [review]

**Comment by theo-3** [comment]
I improved the apply name proc. It uses initial(name) and checks if 'experimental' was there, and also reapplies labels. This also means it won't get confused if you had a label like '([][])' present before setting a variant, which would have messed up the name before.

**Comment by Wojz** [comment]
Good idea.

**Comment by cluwneythescourge** [comment]
man, I've been waiting for something like this forever.  Does the modified name show on the actual plant?   Like, if you grow a "healing" tomato seed, are the plants labeled?

**Comment by theo-3** [comment]
The actual tomatoes are not labeled, since I thought it'd be weird to see a label on plants like that, but I could make it show up there too

**Comment by theo-3** [comment]
There, I added the label to growns too

**Comment by theo-3** [comment]
They're separate in plant bags now too, just tested it

**Comment by theo-3** [comment]
How can I tell whether a player pressed cancel or entered an empty string? Right now pressing cancel deletes the variant instead of doing nothing

**Comment by dearmochi** [comment]
> How can I tell whether a player pressed cancel or entered an empty string? Right now pressing cancel deletes the variant instead of doing nothing

Check if the result of the input is `null` or otherwise empty.

**Comment by theo-3** [comment]
Yeah, I tried using !V, but that prevents a player from deleting a variant by entering an empty string. How do I tell the difference between canceling and typing an empty string?

**Comment by dearmochi** [comment]
Cancelling specifically returns `null` while an empty string is, well, an empty string.

```dm
if(isnull(V))
	// Cancelled
else if(!length(V))
	// Empty string
```

**Comment by theo-3** [comment]
Huh. There's a difference between isnull(V) and V == null?

**Comment by theo-3** [comment]
`var/V = stripped_input(user, "Choose variant name:", "Plant Variant Naming", variant, 64)
		if((!Adjacent(user)) || isnull(V))
			[return`](url)
This isn't returning, pressing cancel deletes the variant

**Comment by theo-3** [comment]
Capture
This isn't working. Pressing cancel still deletes the variant

**Comment by theo-3** [comment]
Yep, it works now. The weird thing is I look at the pandemic machine and it doesn't do this. It just checks for null text without delaying sanitize.

**Comment by theo-3** [comment]
Nevermind, the pandemic machine just makes unlabeled bottles if you cancel

**Comment by theo-3** [comment]
There, now cancel works right

**Comment by theo-3** [comment]
I moved the variant name prompt to a proc to remove code reuse from the gene modder and added naming to trays too. I'm using an adj parameter to get the proc to call Adjacent on the correct object. Is that the right way to do that?

**Review by Kyep — APPROVED** [review]
While I am fine with the concept, I wonder if this requires TGUI. Including raw HTML seems... not great.

**Comment by theo-3** [comment]
> While I am fine with the concept, I wonder if this requires TGUI. Including raw HTML seems... not great.

All botany machines use raw HTML, don't they?

File: code/modules/hydroponics/grown.dm [hunk 5]
```diff
@@ -42,6 +42,8 @@
 		seed.prepare_result(src)
 		transform *= TRANSFORM_USING_VARIABLE(seed.potency, 100) + 0.5 //Makes the resulting produce's sprite larger or smaller based on potency!
 		add_juice()
+		if(seed.variant && seed.variant != "")
```
Comment (AffectedArc07) [inline code/modules/hydroponics/grown.dm:45]: A string of `""` evalutes as false.
image
The second check here (and all other places) is redundant.

**Review by AffectedArc07 — CHANGES_REQUESTED** [review]

**Comment by AffectedArc07** [comment]
> While I am fine with the concept, I wonder if this requires TGUI. Including raw HTML seems... not great.

I am not going to enforce a TGUI conversion in this PR since it is outside the scope.

File: code/modules/hydroponics/gene_modder.dm [hunk 6]
```diff
@@ -391,6 +398,15 @@
...
 		operation = ""
 		target = null
+	else if(href_list["set_v"])
+		if(!seed)
+			return
+		seed.variant_prompt(usr, src)
+	else if(href_list["del_v"])
+		if(!seed)
+			return
+		seed.variant = ""
```
[the remark this answers is not in the document]
Comment (theo-3) [reply code/modules/hydroponics/gene_modder.dm]: Alright, I replaced all empty strings with null. Also fixed a bunch of == "" that wouldn't find null

File: code/modules/hydroponics/grown.dm (hunk 5 above)
Comment (theo-3) [reply code/modules/hydroponics/grown.dm]: Done

**Comment by theo-3** [comment]
Uh. How did this break the tests?

**Comment by theo-3** [comment]
Is this because I force-changed the commit message?

**Comment by theo-3** [comment]
Let's try this again

**Comment by theo-3** [comment]
Alright, I tested it locally and it seems to work. @AffectedArc07 Is it good now?

**Comment by theo-3** [comment]
Quick question: is the stripped_input function supposed to convert nulls to empty strings? It makes it impossible to tell the difference between canceling and empty strings. I tested this with the pandemic machine, which seems to check for cancel too (`if(name == null || wait)`), and canceling creates a bottle named 'culture bottle' instead of doing nothing.

**Review by variableundefined — APPROVED** [review]

**Pull request merged**